backport upgrade.yml changes to BDR-Always-ON

RichardPooleEDB opened the pull request.

amenonsen committed:
Reuse the pgd_proxy/pgdcli/pgbouncer upgrade.yml in BDRAO upgrade.yml

amenonsen committed:
Introduce etcd:upgrade.yml and use it in BDRAO upgrade.yml

amenonsen committed:
Fence/unfence the node using harpctl

This just reuses the code written for the PGD-Always-ON upgrade. Since
failover_manager is set to harp here, it will always use harpctl.

Also introduce harp:upgrade.yml and use that to upgrade both
harp-manager and harp-proxy after the node is fenced.

JonathanRenon-EDB commented:
for major upgrade from 3 to 4 you might have an issue with pgd-cli config file not being present, you will install everything except the config file with the upgrade .yml

Why do we need pgd-proxy steps, it's not upgrading to pgd5 here (BDR-Always-ON/commands/upgrade.yml) since it's done via pgd-always-on/commands/upgrade.yml

RichardPooleEDB committed:
Check versions when upgrading BDR-Always-ON

When upgrading a BDR-Always-ON cluster, check that the BDR major version
currently in use matches the one in config.yml; the logic to do this is
backported from similar check in the PGD-Always-ON version of
upgrade.yml .

Remove references to pgd-proxy from the upgrade process, as it's
pgd5-only so a BDR-Always-ON cluster can never have it.

RichardPooleEDB commented:
The latest commit removes the pgd-proxy upgrade tasks, as it's never used in a BDR-Always-ON cluster. As we've decided not to support upgrading between 3 and 4, we're not going to have a problem with the pgd-cli config. I think this PR is ready now - I've tested that it can upgrade an old 3.7 to latest 3.7, and an old 4.3 to latest 4.3 (and that it will refuse to upgrade a 3.7 to 4).

amenonsen reviewed (approved):
The overall shape looks OK to me, but I haven't run the upgrade myself. @RichardPooleEDB since it's working for you, feel free to merge.
